### ISSUE TITLE
VIMTempFileMakerErrorDomain Code=-11807 being called with ample disk space

### ISSUE THREAD

**Issue by ukudala** [vimeo/VIMNetworking#127]
@alfiehanssen 

##### Overview

`VIMTempFileMakerErrorDomain Code=-11807 "Operation Stopped"` is being sent when ample disk space is left. I observe this code being called following iOS low disk space alert (below).

As it currently stands, to give users a fighting chance of finishing the file copy procedure, utilization of VIMNetworking would require users to have:

> `availableDiskSpace` ==  `fileSize` + .55(`fileSize`) 

of free space in order for fileCopy to complete without the copy operation being cancelled.

55% of fileSize, in addition to original file size, was arrived upon via testing explained below.

![fullsizerender 4](https://cloud.githubusercontent.com/assets/10834575/9619513/05b2435e-50ce-11e5-89f8-426a8602b35f.jpg)

##### Testing procedure

Testing this required utilizing an album in Photos synced via iTunes in order to transfer video files to iOS device (iPad Air). In order to take up disk space, created a number of video files with differing sizes, from 200MB to 1.8GB. Could then mix and match the files, with duplicates, to arrive at differing levels of available disk space. Then able to select from a file size that would make a good test.

For instance, if wishing to test that iOS would throw a destructive low-disk space alert on any system write activity in which file upload size was >.55% of available disk space, can adjust the available disk space on the device and select a file for upload (which will require file copy) in accordance with this test.

> Testing note: TMP directory is cleared on run (not on foregrounding from background). This allows for accurate device disk-space-loading by killing the app, running the app to clear TMP, then proceeding with transferring via iTunes.

------------

##### Testing results:
<img width="532" alt="screen shot 2015-09-01 at 5 09 48 pm" src="https://cloud.githubusercontent.com/assets/10834575/9619342/6b5177a4-50cc-11e5-990c-9ffaadde4a8e.png">

The failed test with green table cells does a nice job of showing this bug, here is a video of the full test:
https://vimeo.com/user37760602/review/138020685/b2f3bfb204

######  Console readout, `green test`:
```
2015-09-01 16:30:22.370 AppTitle[4927:939086] --UPLOAD started
2015-09-01 16:30:22.371 AppTitle[4927:939086] --CREATE started
2015-09-01 16:30:22.386 AppTitle[4927:939086] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:30:22.388 AppTitle[4927:939018] NSManagedObjects did change.
2015-09-01 16:30:23.656 AppTitle[4927:939104] --sessionTaskDidComplete
2015-09-01 16:30:23.656 AppTitle[4927:939104] --COPY started
2015-09-01 16:30:24.164 AppTitle[4927:939018] NSManagedContext did save.
2015-09-01 16:30:37.390 AppTitle[4927:939018] NSManagedContext did save.
2015-09-01 16:33:23.491 AppTitle[4927:939709] --COPY failed
2015-09-01 16:33:23.492 AppTitle[4927:939709] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:33:23.502 AppTitle[4927:939709] --CREATE failed Error Domain=VIMTempFileMakerErrorDomain Code=-11807 "Operation Stopped" UserInfo=0x17046a7c0 {NSUnderlyingError=0x174441320 "The operation couldn’t be completed. No space left on device", NSLocalizedRecoverySuggestion=There is not enough available space to continue the file writing. Make room by deleting existing videos or photos., NSLocalizedDescription=Operation Stopped}
2015-09-01 16:33:23.525 AppTitle[4927:939709] Operation Stopped
2015-09-01 16:33:23.525 AppTitle[4927:939709] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:33:23.532 AppTitle[4927:939709] --UPLOAD failed Error Domain=VIMTempFileMakerErrorDomain Code=-11807 "Operation Stopped" UserInfo=0x17046a7c0 {NSUnderlyingError=0x174441320 "The operation couldn’t be completed. No space left on device", NSLocalizedRecoverySuggestion=There is not enough available space to continue the file writing. Make room by deleting existing videos or photos., NSLocalizedDescription=Operation Stopped}
2015-09-01 16:33:23.534 AppTitle[4927:939709] --NEXT 0: queue complete!
2015-09-01 16:33:23.537 AppTitle[4927:939714] 0
2015-09-01 16:33:23.538 AppTitle[4927:939714] Film not found
```

------------------

On the contrary, the successful test marked with pink cells shows that even with iOS displaying this alert (at the *very* end), the upload can complete, here is a view of that:
https://vimeo.com/user37760602/review/138028445/5181de2d5f

######  Console readout, `pink test`:
```
}
2015-09-01 16:49:17.592 AppTitle[4985:945931] NSManagedObjects did change.
2015-09-01 16:49:17.600 AppTitle[4985:945931] upcoming?
2015-09-01 16:49:19.634 AppTitle[4985:945931] NSManagedContext did save.
2015-09-01 16:49:27.176 AppTitle[4985:945931] POST: priv/video -> gameId : 5270040567873536, offset : 0, recordingStart : 1440613736, recordingDuration : 3961, assetName : Full - 1.82 GB.mp4, cameraAngle : Press box, verified : false
2015-09-01 16:49:27.201 AppTitle[4985:945931] NSManagedObjects did change.
2015-09-01 16:49:27.510 AppTitle[4985:945931] RESPONSE: https://AppTitleio.appspot.com/priv/video -> gameId : 5270040567873536, offset : 0, recordingStart : 1440613736, recordingDuration : 3961, assetName : Full - 1.82 GB.mp4, cameraAngle : Press box, verified : false ---> {
    srvId = 5057548403081216;
    success = 1;
}
2015-09-01 16:49:27.513 AppTitle[4985:945989] SAVING: {
    tasks =     (
        "<VIMUploadTask: 0x17015e8d0>"
    );
}
2015-09-01 16:49:27.513 AppTitle[4985:945989] --NEXT
2015-09-01 16:49:27.514 AppTitle[4985:945989] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:49:27.514 AppTitle[4985:945989] --UPLOAD started
2015-09-01 16:49:27.516 AppTitle[4985:945989] --CREATE started
2015-09-01 16:49:27.533 AppTitle[4985:945989] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:49:27.535 AppTitle[4985:945931] NSManagedObjects did change.
2015-09-01 16:49:28.379 AppTitle[4985:945972] --sessionTaskDidComplete
2015-09-01 16:49:28.379 AppTitle[4985:945972] --COPY started
2015-09-01 16:49:29.191 AppTitle[4985:945931] NSManagedContext did save.
2015-09-01 16:49:42.537 AppTitle[4985:945931] NSManagedContext did save.
2015-09-01 16:52:51.421 AppTitle[4985:946733] --COPY succeeded
2015-09-01 16:52:51.424 AppTitle[4985:946733] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:52:51.424 AppTitle[4985:946733] --CREATE succeeded
2015-09-01 16:52:51.426 AppTitle[4985:946733] --FILE_UPLOAD started
2015-09-01 16:52:51.436 AppTitle[4985:946733] SAVING: {
    tasks =     (
    );
}
2015-09-01 16:52:52.029 AppTitle[4985:945931] 0
2015-09-01 16:52:52.029 AppTitle[4985:945931] Film not found
2015-09-01 16:52:52.030 AppTitle[4985:945931] 1.80547378581297e-05
2015-09-01 16:52:52.030 AppTitle[4985:945931] 3.61094757162594e-05
2015-09-01 16:52:52.032 AppTitle[4985:945931] 5.416421357438909e-05
2015-09-01 16:52:52.032 AppTitle[4985:945931] 7.221895143251879e-05
2015-09-01 16:52:52.186 AppTitle[4985:945931] 9.027368929064848e-05
2015-09-01 16:52:52.490 AppTitle[4985:945931] 0.0001083284271487782
2015-09-01 16:52:52.572 AppTitle[4985:945931] 0.0001263831650069079
2015-09-01 16:52:52.649 AppTitle[4985:945931] 0.0001444379028650376
2015-09-01 16:52:52.709 AppTitle[4985:945931] 0.0001624926407231673
```

##### Observations

It seems that if you have disk space for the file to copy, plus ~55% of the file-being-copied size, you will not get the error `VIMTempFileMakerErrorDomain Code=-11807 "Operation Stopped"`. It would look like this in practice:

> fileSize: `3.2 GB`
additional space required: .55(3.2)= `1.76 GB`
`availableDiskSpaceRequired` == `4.96 GB`

That's really way too much to require. If VIMNetworking is listening to the iOS system call for low disk space, and piggybacking `VIMTempFileMakerErrorDomain Code=-11807` on this call, it is a very bad thing to do. The copy could most definitely continue forward.

---------------------------
> I acknowledge that my tests were done with rounding to the tenth GB, and thus are not as precise as could have been had I employed writing free spaces and file sizes to the console, and copying these values over.

**Comment by ukudala** [comment]
Updated numbers:

- Turned on logging of file sizes and noticed that a `1.82 GB` file was reduced to `1.69 GB` after sync via iTunes
- Equates to `.9285` of the file

Re-ran the numbers using this assumption of lower the file sizes:

<img width="529" alt="screen shot 2015-09-03 at 2 24 44 pm" src="https://cloud.githubusercontent.com/assets/10834575/9670018/9bb13b24-5247-11e5-9d74-834bbdcf2c7d.png">

Currently requiring user have free space exceeding or equal to the file size and 51% (roughly file-and-a-half) as a an interim solution to avoid receiving `Code=-11807` on file copy

**Comment by alfiehanssen** [comment]
@ukudala we're working on a v2.0 of the upload library. A Swift implementation that will be released later this year. It will remove the copy/export phase from the upload flow. Placing the onus on the development to handle copy / export upfront and pass the local url to the copied / exported file to the library for upload. Hopefully this will address the issue and give you greater control over the process.

**Comment by ukudala** [comment]
@alfiehanssen - how long until v1 is deprecated following release of 2.0?

**Comment by alfiehanssen** [comment]
@ukudala Open to your input, the repo will remain open source and available, will just put most effort into v2.0 going forward. Do you have specific concerns?

**Comment by alfiehanssen** [comment]
Also, @ukudala, let's move discussion here: https://github.com/vimeo/VIMUpload/issues/11
